### PR TITLE
19759 - Partial refunds PAYBC implementation

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #:*
https://github.com/bcgov/entity/issues/19759

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 1]
```diff
@@ -45,6 +51,34 @@
...
+    linenumber: str
+    revenueaccount: str
+    revenueamount: Decimal
+    glstatus: str
+    glerrormessage: Optional[str]
+    refund_data: List[RefundData]
+
+
+@define
+class OrderStatus():
```
Comment (seeker25) [inline pay-api/src/pay_api/services/direct_pay_service.py:76]: I'll refactor these out of the payment job direct_pay automated refunds

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 2]
```diff
@@ -234,19 +269,85 @@ def _get_refund_token(cls):
...
+        access_token: str = DirectPayService().get_token().json().get('access_token')
+        paybc_ref_number: str = current_app.config.get('PAYBC_DIRECT_PAY_REF_NUMBER')
+        paybc_svc_base_url = current_app.config.get('PAYBC_DIRECT_PAY_BASE_URL')
+        completed_reference = list(
+            filter(lambda reference: (reference.status_code == InvoiceReferenceStatus.COMPLETED.value),
+                   invoice.references))[0]
+        payment_url: str = \
+            f'{paybc_svc_base_url}/paybc/payment/{paybc_ref_number}/{completed_reference.invoice_number}'
+        payment_response = OAuthService.get(payment_url, access_token, AuthHeaderType.BEARER, ContentType.JSON).json()
+        return Converter().structure(payment_response, OrderStatus)
```
Comment (seeker25) [inline pay-api/src/pay_api/services/direct_pay_service.py:330]: straight forward to serialize / deserialize, will look closer tomorrow

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 3]
```diff
@@ -234,19 +269,85 @@ def _get_refund_token(cls):
...
+                revenue_account = DirectPayService._get_gl_coding(service_fee_distribution, refund.refund_amount)
+            refund_lines.append({
+                'revenueaccount': revenue_account,
+                'refundAmount': refund.amount
+            })
+            total += refund.amount
+        return refund_lines, total
+
+    @staticmethod
+    def _query_order_status(invoice: InvoiceModel) -> OrderStatus:
```
Comment (seeker25) [inline pay-api/src/pay_api/services/direct_pay_service.py:319]: Will refactor DIRECT_PAY to reuse this hopefully, not sure about the other job though... we'll see

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 4]
```diff
@@ -234,19 +269,85 @@ def _get_refund_token(cls):
...
+        """Request invoice order status from PAYBC."""
+        access_token: str = DirectPayService().get_token().json().get('access_token')
+        paybc_ref_number: str = current_app.config.get('PAYBC_DIRECT_PAY_REF_NUMBER')
+        paybc_svc_base_url = current_app.config.get('PAYBC_DIRECT_PAY_BASE_URL')
+        completed_reference = list(
```
Comment (seeker25) [inline pay-api/src/pay_api/services/direct_pay_service.py:324]: I think there should only be one `invoice_reference` that's completed, can't remember why this is here unfortunately

File: pay-api/setup.cfg [hunk 5]
```diff
@@ -71,7 +71,7 @@ notes=FIXME,XXX,TODO
 ignored-modules=flask_sqlalchemy,sqlalchemy,SQLAlchemy,alembic,scoped_session
 ignored-classes=scoped_session
 min-similarity-lines=15
-disable=C0301,W0511
+disable=C0301,W0511,R0903
```
Comment (seeker25) [inline pay-api/setup.cfg:74]: Too few public methods warning, it's fairly useless for dataclasses etc

**Comment by seeker25** [comment]
Creating another ticket to do refactoring of PAYBC job

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 6]
```diff
@@ -234,19 +281,92 @@ def _get_refund_token(cls):
...
             'pbcRefNumber': current_app.config.get('PAYBC_DIRECT_PAY_REF_NUMBER'),
             'txnAmount': invoice.total,
             'txnNumber': invoice_reference.invoice_number
         }
+        if not refund_partial:
+            return refund_payload
+
+        refund_lines, total_refund = DirectPayService._build_refund_revenue_lines(refund_partial)
+        paybc_invoice = DirectPayService._query_order_status(invoice)
+        refund_payload.update({
```
Comment (seeker25) [inline pay-api/src/pay_api/services/direct_pay_service.py:368]: I probably could have used a `dataclass` for this, but I think i'll leave it for now

**Comment by seeker25** [comment]
Haven't tested with PAYBC yet, standby

File: pay-api/tests/unit/services/test_direct_pay_service.py [hunk 7]
```diff
@@ -257,5 +260,172 @@ def test_process_cfs_refund_duplicate_refund(monkeypatch):
...
+    """Assert validations are working correctly for building refund payload."""
+    invoice, payment_line_item = _automated_refund_preparation()
+    if test_name == 'pay_pli_not_exist':
+        refund_partial[0].payment_line_item_id = 999999999
+    else:
+        refund_partial[0].payment_line_item_id = payment_line_item.id
+    direct_pay_service = DirectPayService()
+    with pytest.raises(BusinessException) as excinfo:
+        direct_pay_service.build_automated_refund_payload(invoice, refund_partial)
+        assert excinfo.value.code == Error.INVALID_REQUEST.name
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_direct_pay_service.py:338]: I wanted to return real errors, not sure your thoughts about returning INVALID REQUEST, on the plus side at least I wrote some logging to handle the cases

Comment (ochiu) [reply pay-api/tests/unit/services/test_direct_pay_service.py:338]: This works for me

**Review by ochiu — APPROVED** [review]

**Pull request merged**